manticore/platforms/evm.py

```diff
@@ -0,0 +1,1760 @@
+        #    raise NotEnoughGas()
+        self.memory.allocate(address)
+
+    def _store(self, address, value):
+        #CHECK ADDRESS IS A 256 BIT INT OR BITVEC
+        #CHECK VALUE IS A 256 BIT INT OR BITVEC
+        if address > 100000:
+            raise NotEnoughGas()
+        self._allocate(address)
+        self.memory.write(address, [value])
+
+
+    def _load(self, address):
+        self._allocate(address+32)
+        value = self.memory.read(address,1)[0]
+        #print pretty_print(value)
+        value = arithmetic_simplifier(value)
+        if isinstance(value, Constant) and not value.taint: 
+            value = value.value
+        return value
+
+    @staticmethod
+    def check256int(value):
+        assert True
+
+    def read_code(self, address, size=1):
+        '''
+            Read size byte from bytecode.
+            If less than size bytes are available result will be pad with \x00
+        '''
+        assert address < len(self.bytecode)
+        value = self.bytecode[address:address+size]
+        if len(value) < size:
+            value += '\x00'* (size-len(value)) #pad with null (spec)
+        return value
+
+    def disassemble(self):
+        return EVMDecoder.disassemble(self.bytecode)
+
+
+    @property
+    def PC(self):
+        return self.pc
+
+    @property
+    def instruction(self):
+        ''' 
+            Current instruction pointed by self.pc 
+        '''
+        #FIXME check if pc points to invalid instruction
+        #if self.pc >= len(self.bytecode):
+        #    return InvalidOpcode('Code out of range')
+        #if self.pc in self.invalid:
+        #    raise InvalidOpcode('Opcode inside a PUSH immediate')
+
+        def getcode():
+            for byte in self.bytecode[self.pc:]:
+                yield byte
+            while True:
+                yield '\x00'
+
+        return EVMDecoder.decode_one(getcode())
+
+    #auxiliar funcs
+    #Stack related
+    def _push(self, value):
+        '''
+                   ITEM0
+                   ITEM1
+                   ITEM2
+             sp->  {empty}
+        '''
+        assert isinstance(value,(int,long)) or isinstance(value, BitVec) and value.size==256 
+        if len(self.stack) >= 1024:
+            raise StackOverflow()
+        self.stack.append(value & TT256M1)
+
+    def _pop(self):
+        if len(self.stack) == 0:
+            raise StackUnderflow()
+        return self.stack.pop()
+        
+
+    #Execute an instruction from current pc
+    def execute(self):
+        if issymbolic(self.pc):
+            expression = self.pc
+            def setstate(state, value):
+                state.platform.current.pc = value
+            raise Concretize("Concretice PC",
+                                expression=expression, 
+                                setstate=setstate,
+                                policy='ALL')
+
+        self.publish('will_decode_instruction', self.pc)
+        current = self.instruction
```

Comment:
Should we publish `did_decode_instruction` here and `did_execute_instruction` below?

Reply:
We don't have that. Maybe we should. Though, in that case we need to upgrade abstract cpu et.at. I think that's for a different PR.